src/vivarium_public_health/metrics/new_disability.py

```diff
@@ -0,0 +1,102 @@
+"""
+===================
+Disability Observer
+===================
+
+This module contains tools for observing years lived with disability (YLDs)
+in the simulation.
+
+"""
+import typing
+
+import pandas as pd
+from vivarium.framework.values import list_combiner, union_post_processor, rescale_post_processor
+
+from vivarium_public_health.disease import DiseaseState, RiskAttributableDisease
+from vivarium_public_health.utilities import to_years
+
+if typing.TYPE_CHECKING:
+    from vivarium.framework.engine import Builder
+
+# I'm trying to think about how we make these things easier to subclass.
+# Perhaps if we are careful we can avoid subclassing by being clever in metrics
+# but I think we'll run into too many special cases.
+#
+# I'm thinking that maybe we can design inheritance simply around overriding
+# properties.
+
+
+class DisabilityObserver:
+
+    @property
+    def name(self):
+        return 'disability_observer'
+
+    @property
+    def disease_classes(self):
+        return [DiseaseState, RiskAttributableDisease]
+
+    @property
+    def additional_groupers(self):
+        return []
+
+    @property
+    def additional_columns(self):
+        return []
+
+    def setup(self, builder: 'Builder'):
+        self.disability_weight = builder.value.register_value_producer(
+            'disability_weight',
+            source=lambda index: [pd.Series(0.0, index=index)],
+            preferred_combiner=list_combiner,
+            preferred_post_processor=_disability_post_processor)
+
+        cause_states = builder.components.get_components_by_type(tuple(self.disease_classes))
+        cause_names = [c.state_id for c in cause_states]
+        self.disability_weight_pipelines = {cause: builder.value.get_value(f'{cause}.disability_weight')
+                                            for cause in cause_names}
+
+        params = builder.configuration.metrics.stratification_params
+        groupers = [] + self.additional_groupers
+        columns = ['tracked', 'alive'] + self.additional_columns
+        if 'age_group' in params:
```

Comment:
Still thinking about how to get rid of this block; it'll show up in every observer for now.

Reply:
The issue is mainly around year, which requires a different grouping for different observers (e.g. exit year for deaths)

Reply:
Actually, strike that.  we'll probably just collect deaths each time step instead.